File: emit.py

class Emitter:
    def __init__(self, fullPath):
        self.fullPath = fullPath
        self.header = ""
        self.code = ""
        self.lvl = 0

    def increaseLvl(self):
        self.lvl+=1

    def decreaseLvl(self):
        self.lvl-=1
        self.lvl = max(self.lvl, 0)

    def emit(self, code, push=False):
        if push:
            for i in range (self.lvl):
                self.code+="\t"
        self.code += code

    def emitLine(self, code, push=False):
        if push:
            for i in range (self.lvl):
                self.code+="\t"
        self.code += code + '\n'

    def headerLine(self, code):
        self.header += code + '\n'

    def writeFile(self):
        with open(self.fullPath, 'w') as outputFile:
            outputFile.write(self.header + self.code)

File: cpp/emit.cpp
#include "emit.h"

Emitter::Emitter(const string& fullPath) : fullPath(fullPath), lvl(0)
{
    header.reserve(1000);
    code.reserve(1000);
}

inline void Emitter::increase_level() {
    lvl++;
}

inline void Emitter::decrease_lvl() {
    lvl--;
    lvl = max(0, lvl);
}

void Emitter::emit(const string& code, bool push) {
    if (push) {
        for (int i = 0; i < lvl; ++i) {
            this->code+= "\t";
        }
    }
    this->code += code;
}

void Emitter::emit_line(const string&code, bool push) {
    if (push) {
        for (int i = 0; i < lvl; ++i) {
            this->code+= "\t";
        }
    }
    this->code += code;
    this->code += "\n";
}

void Emitter::header_line(const string& code, bool push) {
    if (push) {
        this->header += "\t";
    }
    this->header += code;
    this->header += "\n";
}

void Emitter::write_file() {
    ofstream output (fullPath);
    if (!output.is_open()) {
        cerr << "Failed to open file\n";
        return;
    }

    output << "Hello, world!\n";
    output.close();
}